File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/intake/Rotate.java
package org.firstinspires.ftc.teamcode.intake;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.Disabled;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.systems.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.systems.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Intake.Rotate.pickup;
import static org.firstinspires.ftc.teamcode.Positions.Intake.Rotate.transfer;

/**
 * Class representing the vertical rotation of the intake claw.
 * It extends the ServoPositionMechanism class to control the rotation servo.
 */
public class Rotate extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servo from.
     */
    public Rotate(HardwareMap hardwareMap) {
        super(transfer);
        this.hardwareMap = hardwareMap;
    }
    /**
     * Gets the servo associated with the Rotate mechanism.
     * @return an array containing the rotation servo.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Intake Rotate")};
    }
}

/**
 * TeleOp class for testing the Rotate mechanism manually.
 */
@TeleOp(name = "Intake Rotate Test", group = "C Intake")
@Disabled
class RotateTest extends ManualMechanismTeleOp {
    public RotateTest() {
        super(Rotate::new);
    }
}

/**
 * TeleOp class for testing the Rotate mechanism positions.
 * @noinspection DuplicatedCode
*/
@TeleOp(name = "Intake Rotate Positions Test", group = "D Intake")
@Config
@Disabled
class RotatePositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Rotate rotate = new Rotate(hardwareMap);

        waitForStart();

        while (opModeIsActive()) {
            if (gamepad1.dpad_up) current = pickup;
            else if (gamepad1.dpad_down) current = transfer;
            else current += rotate.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            rotate.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if (!rotate.run(packet)) requestOpModeStop();
            FtcDashboard.getInstance().sendTelemetryPacket(packet);
            telemetry.addData("Current Position", rotate.getTargetPosition());
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/intake/Rotate.kt
package org.firstinspires.ftc.intake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.intake.Rotate.Companion.pickup
import org.firstinspires.ftc.intake.Rotate.Companion.pickupWait
import org.firstinspires.ftc.intake.Rotate.Companion.transfer
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.ServoPositionMechanism

@Config
class Rotate(hardwareMap: HardwareMap) : ServoPositionMechanism(transfer) {
    companion object {
        /**
         * The position of the intake rotation when it is outside of the submersible.
         */
        @JvmField
        @Volatile
        var pickupWait = 0.1
        /**
         * The position of the intake rotation when it is inside of the submersible.
         */
        @JvmField
        @Volatile
        var pickup = 0.05
        /**
         * The position of the intake rotation when it is in the transfer position.
         */
        @JvmField
        @Volatile
        var transfer = 0.75
    }
    /**
     * Gets the servo associated with the Rotate mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Intake Rotate"])
}
/**
 * TeleOp class for testing the Rotate mechanism manually.
 */
@TeleOp(name = "Intake Rotate Test", group = "C")
@Disabled
private class RotateTest : ManualMechanismTeleOp(::Rotate)
/**
 * TeleOp class for testing the positions of the Rotate mechanism.
 */
@TeleOp(name = "Intake Rotate Position Test", group = "D")
@Disabled
private class RotatePositions : Movement() {
    /**
     * The Rotate mechanism to be tested.
     */
    override lateinit var system: Rotate

    override fun systemInit() {
        system = Rotate(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = pickupWait })
        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = pickup })
        actions.add(PressAction(gamepad1::dpad_up) { system.targetPosition = transfer })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}
